src/Validation/RulesGatherer.php

```diff
@@ -175,20 +191,53 @@ protected function wrap(array $rulesOrMessages, array $path): array
         $withPath = [];
 
         foreach ($rulesOrMessages as $key => $value) {
-            $combinedPath = array_merge($path, [$key]);
-            $pathDotNotation = $this->pathDotNotation($combinedPath);
+            $combinedPath = implode('.', array_merge($path, [$key]));
 
-            $withPath[$pathDotNotation] = $value;
+            $withPath[$combinedPath] = $value;
         }
 
         return $withPath;
     }
 
     /**
-     * @param  array<int|string>  $path
+     * Prepend rule arguments that refer to other arguments with the full path.
+     *
+     * This may be necessary to allow certain rules to be reusable when placed
+     * upon input arguments. For example, `required_with:foo` may be defined
+     * on an input value that is nested within the arguments under `input.0`.
+     * It is thus changed to the full reference `required_with:input.0.foo`.
+     *
+     * @param  array<int, mixed>  $rules
+     * @param  array<int|string>  $argumentPath
+     * @return array<int, array<int, mixed>|\Illuminate\Contracts\Validation\Rule>
      */
-    protected function pathDotNotation(array $path): string
+    protected function qualifyArgumentReferences(array $rules, array $argumentPath): array
     {
-        return implode('.', $path);
+        return array_map(
+            static function ($rule) use ($argumentPath) {
+                if ($rule instanceof Rule) {
+                    return $rule;
+                }
+
+                /**
+                 * @var array{
+                 *   0: string,
+                 *   1: array<int, mixed>,
+                 * } $parsed
+                 */
+                $parsed = ValidationRuleParser::parse($rule);
+
+                $name = $parsed[0];
+                if ($name === 'RequiredWithout') {
```

Comment:
Is there a clean way of making this more abstract to work with any validator of this format? 
e.g. `RequiredWith`, `RequiredWithAll`, `RequiredWithoutAll`, `RequiredUnless`, `Same`

Reply:
We can treat all rules that have the argument reference as the first argument in the same way. Are there any rules that also reference args but in another position, e.g. `some_rule:3,arg_name`?

If you could dig in and provide a comprehensive list, that would be of great help :)

Reply:
[This array](https://github.com/illuminate/validation/blob/master/Validator.php#L198-L225) gives us a complete list of validation rules which depend on other fields as parameters.  However, it both lists validation rules which are dependant on scalar values (e.g. `after:date`)  as references to other attributes (e.g. `required_with`)

Combining this information with the [laravel docs on validation](https://laravel.com/docs/8.x/validation), we can extract the following complete list of validation rules where one of the arguments are references to other fields.
```
    /**
     * The validation rules which depend on other attributes as parameters.
     *
     * @var string[]
     */
    protected $dependentRules = [
        'Different',
        'ExcludeIf',
        'ExcludeUnless',
        'ExcludeWithout',
        'Gt',
        'Gte',
        'Lt',
        'Lte',
        'RequiredIf',
        'RequiredUnless',
        'RequiredWith',
        'RequiredWithAll',
        'RequiredWithout',
        'RequiredWithoutAll',
        'Same'
    ];
```

All of these validation rules take `field` or `fields` as their first parameter. `exclude_if` and `exclude_unless` take a value as second argument.